libsql-remote/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+    mod cloudflare;
+
+    pub type DbConnection = crate::cloudflare::DbConnection;
+}
+
+pub mod connection;
+pub mod params;
+pub mod pipeline;
+pub mod proto;
+
+pub use params::{IntoParams, Params};
+pub use pipeline::{
+    ClientMsg, Response, ServerMsg, StreamBatchReq, StreamExecuteReq, StreamRequest,
+    StreamResponse, StreamResponseError, StreamResponseOk,
+};
+pub use proto::Value;
+use proto::{Col, Stmt, StmtResult};
+
+use crate::connection::Connection;
+use base64::engine::general_purpose::STANDARD_NO_PAD;
+use base64::Engine;
+use serde::ser::{SerializeSeq, SerializeStruct};
+use serde::{Serialize, Serializer};
+use serde_json::json;
+use std::collections::VecDeque;
+use std::future::Future;
+use std::ops::{Deref, Index};
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+
+type Result<T> = std::result::Result<T, Error>;
+
+pub trait HttpSend<'a> {
+    type Result: Future<Output = Result<ServerMsg>> + 'a;
+    fn http_send(&'a self, url: String, auth: String, body: String) -> Self::Result;
+}
+
+pub type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("unexpected response: `{0}`")]
+    UnexpectedResponse(String),
+    #[error("stream closed: `{0}`")]
+    StreamClosed(String),
+    #[error("stream error: `{0:?}`")]
+    StreamError(StreamResponseError),
+    #[error("json error: `{0}`")]
+    Json(#[from] serde_json::Error),
+    #[error("http error: `{0}`")]
+    Http(String),
+    #[error("api error: `{0}`")]
+    Api(String),
+    #[error("SQL conversion failure: `{0}`")]
+    ToSqlConversionFailure(BoxError),
+    #[error("Column not found: {0}")]
+    ColumnNotFound(i32),
+}
+
+pub struct Statement<T> {
+    client: Connection<T>,
+    inner: Stmt,
+}
+
+impl<T> Statement<T>
+where
+    T: for<'a> HttpSend<'a>,
+{
+    pub async fn execute(&mut self, params: &Params) -> Result<usize> {
+        let mut stmt = self.inner.clone();
+        bind_params(params.clone(), &mut stmt);
+
+        let v = self.client.execute_inner(stmt, 0).await?;
+        let affected_row_count = v.affected_row_count as usize;
+        self.client
+            .affected_row_count
+            .store(affected_row_count as u64, Ordering::SeqCst);
+        if let Some(last_insert_rowid) = v.last_insert_rowid {
+            self.client
+                .last_insert_rowid
+                .store(last_insert_rowid, Ordering::SeqCst);
+        }
+        Ok(affected_row_count)
+    }
+
+    pub async fn query(&mut self, params: &Params) -> Result<Rows> {
+        let mut stmt = self.inner.clone();
+        bind_params(params.clone(), &mut stmt);
+
+        let StmtResult { rows, cols, .. } = self.client.execute_inner(stmt, 0).await?;
+
+        Ok(Rows {
+            rows,
+            cols: Arc::new(cols),
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Rows {
+    cols: Arc<Vec<Col>>,
+    rows: VecDeque<Vec<Value>>,
+}
+
+impl Serialize for Rows {
```

Comment:
I don't totally understand why we have a custom Serialize implementation for rows here? Could you expand on why this is needed esp such a custom implementation?

Reply:
proto::Value is serialized for the sake of Hrana protocol using quite elaborate schema (ie. every cell is `{"value":"123"}`). 

Originally this code was put in my Cloudflare worker example app to make JSON data set returned by the API more user friendly, but since it's a lot of manual labor to write it for every user that simply wants to return a SQL result data set as a JSON, I've put it here.